submitit/core/core.py

```diff
@@ -507,6 +515,92 @@ def __setstate__(self, state: tp.Dict[str, tp.Any]) -> None:
+        """ same as wait() but with asyncio sleep.
+        """
+        while not self.job.done():
+            await asyncio.sleep(poll_interval)
+
+    async def result(self, poll_interval: tp.Union[int, float] = 1) -> R:
+        """ asyncio version of the result() method.
+        Wait asynchornously for the result to be available by polling the self.done() method.
+        Parameters
+        ----------
+        poll_interval: int or float
+            how often to check if the result is available, in seconds
+        """
+        await self.wait(poll_interval)
+        return self.job.result()
+
+    async def results(self, poll_interval: tp.Union[int, float] = 1) -> tp.List[R]:
+        """ asyncio version of the results() method.
+
+        Waits asynchornously for ALL the results to be available by polling the self.done() method.
+
+        Parameters
+        ----------
+        poll_interval: int or float
+            how often to check if the result is available, in seconds
+        """
+        await self.wait(poll_interval)
+        # results are ready now
+        return self.job.results()
+
+    def results_as_compteled(self, poll_interval: tp.Union[int, float] = 1) -> tp.Iterator[asyncio.Future]:
+        """awaits for all tasks results concurrently. Note that the order of results is not guaranteed to match the order
+            of the tasks anymore as the earliest task coming back might not be the first one you sent.
+
+            Returns
+            -------
+            an iterable of Awaitables that can be awaited on to get the earliest result available of the remaining tasks.
+
+            Parameters
+            ----------
+            poll_interval: int or float
+                how often to check if the result is available, in seconds
+
+            (see https://docs.python.org/3/library/asyncio-task.html#asyncio.as_completed)
+        """
+        if self.job.num_tasks > 1:
+            yield from asyncio.as_completed(
+                [self.job.task(i).async_job().result(poll_interval) for i in range(self.job.num_tasks)]
+            )
+
+        # there is only one result anyway, let's just use async result
+        yield asyncio.ensure_future(self.result())
+
+    ## proxy everything else so this just looks like a normal job
+    def __getattr__(self, item):
+        if hasattr(self.job, item):
+            return getattr(self.job, item)
+        raise AttributeError(item)
+
+    def __nonzero__(self):
+        return bool(self.job)
+
+    def __str__(self):
+        return str(self.job)
+
+    def __repr__(self):
+        return repr(self.job)
+
+    def __hash__(self):
+        return hash(self.job)
+
+    def __del__(self) -> None:
+        del self.job
+
+    def __getstate__(self) -> tp.Dict[str, tp.Any]:
+        return self.job.__getstate__()
+
+    def __setstate__(self, state: tp.Dict[str, tp.Any]) -> None:
+        return self.job.__setstate__(state)
```

Comment:
I think I wouldn't have bothered with all this, the proxy is helpful for asyncio but I guess one should use the initial job naturally for other operations? but fine either way if you prefer like this.
Also, from the moment `getattr` is defined I guess the typing will be mostly useless, so I am not completely fan of it, but it really depends how you want to use it in the end.